src/HttpRequest.cc

```diff
@@ -714,34 +714,96 @@ UpdateRequestNotes(ConnStateData *csd, HttpRequest &request, NotePairs const &he
 }
 
 void
-HttpRequest::manager(const CbcPointer<ConnStateData> &aMgr, const AccessLogEntryPointer &al)
+HttpRequest::prepareForDownloader(Downloader *aDownloader)
+{
+    header.putStr(Http::HdrType::HOST, url.host());
+    header.putTime(Http::HdrType::DATE, squid_curtime);
+    downloader = aDownloader;
+    makeInternal();
+}
+
+void
+HttpRequest::makeInternal()
 {
-    clientConnectionManager = aMgr;
+    /* Internally created requests cannot have bodies today */
+    content_length = 0;
+    http_ver = Http::ProtocolVersion();
+    internal = true;
+}
 
-    if (!clientConnectionManager.valid())
-        return;
+CbcPointer<ConnStateData> &
+HttpRequest::clientConnectionManager()
+{
+    if (!internal)
+        return masterXaction->clientConnectionManager();
+    static CbcPointer<ConnStateData> noManager;
+    return noManager;
+}
 
-    AnyP::PortCfgPointer port = clientConnectionManager->port;
-    if (port) {
-        myportname = port->name;
-        flags.ignoreCc = port->ignore_cc;
-    }
+static const Ip::Address&
+NoAddr()
+{
+    static Ip::Address addr;
+    if (!addr.isNoAddr())
+        addr.setNoAddr();
+    return addr;
+}
+
+const Ip::Address&
+HttpRequest::clientAddr() const
+{
+    if (internal)
+        return NoAddr();
+    if (clientConnection())
+        return clientConnection()->remote;
+    return client_addr;
+}
+
+void
+HttpRequest::prepareForConnectionlessProtocol(const Ip::Address &fromAddr, const Ip::Address &localAddr)
+{
+    client_addr = fromAddr;
+    my_addr = localAddr;
+}
+
+const Ip::Address&
+HttpRequest::myAddr() const
+{
+    return internal ? NoAddr():
+           masterXaction->clientConnection() ?
+           masterXaction->clientConnection()->local : my_addr;
+}
 
-    if (auto clientConnection = clientConnectionManager->clientConnection) {
-        client_addr = clientConnection->remote; // XXX: remove request->client_addr member.
 #if FOLLOW_X_FORWARDED_FOR
-        // indirect client gets stored here because it is an HTTP header result (from X-Forwarded-For:)
-        // not details about the TCP connection itself
-        indirect_client_addr = clientConnection->remote;
+const Ip::Address&
+HttpRequest::indirectClientAddr() const
+{
+    if (internal)
+        return NoAddr();
+    if (!indirect_client_addr.isEmpty())
+        return indirect_client_addr;
+    return clientAddr();
+}
+
+void
+HttpRequest::ignoreIndirectClientAddr()
+{
+    indirect_client_addr = clientAddr();
```

Comment:
The code does not seem to match the method name -- we are setting (a part of) indirect_client_addr instead of making it unusable/empty. Perhaps we can we use setEmpty() here instead?